File: javascript/05-dsi-calculation.js
/**
 * DSI Calculation — Vocametrix API example (Node.js)
 *
 * Computes the Dysphonia Severity Index (DSI).
 *   DSI >  1.6  → normal
 *   DSI <= 1.6  → possible dysphonia
 *   DSI <  0    → severe dysphonia
 *
 * Upload pattern: POST /api/assignFileId → GET /api/calculate-dsi
 * Auth: X-API-Key header
 *
 * Usage:
 *   node 05-dsi-calculation.js path/to/sustained_vowel.wav
 */

import fs from 'fs';
import FormData from 'form-data';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';

async function assignFileId(audioPath) {
  const form = new FormData();
  form.append('audio', fs.createReadStream(audioPath));
  form.append('email', 'user@example.com');
  const res = await fetch(`${BASE_URL}/api/assignFileId`, {
    method: 'POST',
    headers: { 'X-API-Key': API_KEY, ...form.getHeaders() },
    body: form,
  });
  if (!res.ok) throw new Error(`assignFileId failed: ${await res.text()}`);
  return (await res.json()).fileId;
}

async function calculateDsi(audioPath) {
  const svFileId = await assignFileId(audioPath);
  const res = await fetch(`${BASE_URL}/api/calculate-dsi?svFileId=${svFileId}`, {
    headers: { 'X-API-Key': API_KEY },
  });
  if (!res.ok) throw new Error(`calculate-dsi failed: ${await res.text()}`);
  return res.json();
}

function printResults(results) {
  const dsi = results.DSI ?? 'N/A';
  console.log('\n=== DSI Results ===');
  console.log(`DSI score: ${dsi}`);
  if (typeof dsi === 'number') {
    const label = dsi > 1.6 ? 'Normal' : dsi > 0 ? 'Mild dysphonia' : 'Severe dysphonia';
    console.log(`Classification: ${label}`);
  }
  for (const key of ['F0', 'JITTER', 'SHIMMER', 'MPT', 'HNR', 'INTENSITY_MIN']) {
    if (results[key] !== undefined) console.log(`  ${key}: ${results[key]}`);
  }
}

const audioPath = process.argv[2];
if (!audioPath) {
  console.error('Usage: node 05-dsi-calculation.js <sustained_vowel.wav>');
  process.exit(1);
}
calculateDsi(audioPath).then(printResults).catch(err => { console.error(err.message); process.exit(1); });

File: python/05_dsi_calculation.py
"""
DSI Calculation — Vocametrix API example

Computes the Dysphonia Severity Index (DSI) — a multiparametric index
combining highest frequency, lowest intensity, maximum phonation time,
jitter, and shimmer into a single severity score.

  DSI >  1.6  → normal voice quality
  DSI <= 1.6  → possible dysphonia
  DSI <  0    → severe dysphonia

Upload pattern: POST /api/assignFileId → GET /api/calculate-dsi
Auth: X-API-Key header

Usage:
    python 05_dsi_calculation.py path/to/sustained_vowel.wav
"""

import os
import argparse
import requests
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}


def assign_file_id(audio_path: str) -> str:
    with open(audio_path, 'rb') as f:
        r = requests.post(
            f'{BASE_URL}/api/assignFileId',
            headers=HEADERS,
            files={'audio': f},
        )
    r.raise_for_status()
    return r.json()['fileId']


def calculate_dsi(audio_path: str) -> dict:
    file_id = assign_file_id(audio_path)
    r = requests.get(
        f'{BASE_URL}/api/calculate-dsi',
        headers=HEADERS,
        params={'svFileId': file_id},
    )
    r.raise_for_status()
    return r.json()


def print_results(results: dict):
    dsi = results.get('DSI', 'N/A')
    print('\n=== DSI Results ===')
    print(f"DSI score: {dsi}")
    if isinstance(dsi, (int, float)):
        if dsi > 1.6:
            label = 'Normal'
        elif dsi > 0:
            label = 'Mild dysphonia'
        else:
            label = 'Severe dysphonia'
        print(f"Classification: {label}")

    for key in ('F0', 'JITTER', 'SHIMMER', 'MPT', 'HNR', 'INTENSITY_MIN'):
        if key in results:
            print(f"  {key}: {results[key]}")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='DSI calculation')
    parser.add_argument('audio', help='Path to sustained-vowel WAV')
    args = parser.parse_args()

    results = calculate_dsi(args.audio)
    print_results(results)
